### PR TITLE
PTI-508: Tweak legislation act/reg filter logic and UI.

### PR DESCRIPTION
**Pull request by NickPhura**
- Fix business logic around how filtering acts and regs works, and what happens when you clear either field.
- Improve performance of filtering by wrapping filter functions in debounce.
- Add `Clear Selected` text to top option in acts/regs lists.

### CODE REVIEW

File: angular/projects/common/src/app/legislation-add-edit/legislation-add-edit.component.ts [hunk 1]
```diff
@@ -28,9 +29,14 @@ export class LegislationAddEditComponent implements OnInit {
   public filteredActs: string[] = [];
   public filteredRegulations: string[] = [];
 
+  public debouncedFilterActsPicklist = this.utils.debounced(200, args => this.filterActsPicklist(args));
+  public debouncedFilterRegulationsPicklist = this.utils.debounced(200, args => this.filterRegulationsPicklist(args));
```
[the remark this answers is not in the document]
Comment (NickPhura) [reply angular/projects/common/src/app/legislation-add-edit/legislation-add-edit.component.ts:33]: Just for performance reasons.  If you type in the box 5 characters quickly, all of the logic will get run 5 times.  By debouncing, it should just get run 1-2 times (depending on typing speed).

[the remark this answers is not in the document]
Comment (danieltruong) [reply angular/projects/common/src/app/legislation-add-edit/legislation-add-edit.component.ts:33]: Can we use something other than keyup to run your function?

[the remark this answers is not in the document]
Comment (NickPhura) [reply angular/projects/common/src/app/legislation-add-edit/legislation-add-edit.component.ts:33]: I don't think so, not if we want it to function the way it does, as a live filter as you type.

[the remark this answers is not in the document]
Comment (NickPhura) [reply angular/projects/common/src/app/legislation-add-edit/legislation-add-edit.component.ts:33]: The debouncing isn't absolutely necessary, but I do think it is a worthwhile optimization that cuts down on wasted function calls.

**Comment by NickPhura** [comment]
@danieltruong @marklise Have you two finished your reviews?

**Comment by danieltruong** [comment]
Found an issue.

- add an order
- click on either act or regulation
- the list of possible selections are empty
- click something else and then click either act or regulation
- the list is now populated

**Comment by NickPhura** [comment]
> 
> 
> Found an issue.
> 
>     * add an order
> 
>     * click on either act or regulation
> 
>     * the list of possible selections are empty
> 
>     * click something else and then click either act or regulation
> 
>     * the list is now populated
> 
> 
> Expected behaviour should be that the list populates upon first activation of the field.

Committed a change to update the ngOnInit, which wasn't fully handling all scenarios.

**Comment by danieltruong** [comment]
Another oddity:

- Add order
- Select Seed Potato Act
- In regulation type out Seed
- Seed Potato Regulation does not show up.

**Comment by NickPhura** [comment]
> 
> 
> Another oddity:
> 
>     * Add order
> 
>     * Select Seed Potato Act
> 
>     * In regulation type out Seed
> 
>     * Seed Potato Regulation does not show up.

I can't seem to re-produce this.

Steps I tried:
- Add order
- Select 'Seed Potato Act'
- Click Regulation dropdown
  - dropdown shows 3 regulations
- type out 'Seed'
  - The regulation dropdown continues to show 3 regulations  as each of the 3 regulations contains the word seed, so further filtering doesn't occur.

**Comment by NickPhura** [comment]
> 
> 
> Another oddity:
> 
>     * Add order
> 
>     * Select Seed Potato Act
> 
>     * In regulation type out Seed and hit enter
> 
>     * Seed Potato Regulation does not show up.

Pushed an updated to better handle ENTER clicks, as we talked about.  Also did some minor cleanup in a few spots.

**Pull request merged**